posts/2020-05-15-five-years-of-rust.md

```diff
@@ -0,0 +1,508 @@
+with Alternative Registries your organisation can easily build and share your
+own registry of crates that can be used in your projects like they were
+on crates.io.
+
+**[1.39] — Async/Await** The stabilisation of the async/await keywords for
+handling Futures was one of the major milestones to making async programming
+in Rust a first class citizen. Even just six months after its release the
+async programming has blossomed into a diverse and performant ecosystem.
+
+[1.2]: https://blog.rust-lang.org/2015/08/06/Rust-1.2.html
+[1.3]: https://blog.rust-lang.org/2015/09/17/Rust-1.3.html
+[1.4]: https://blog.rust-lang.org/2015/10/29/Rust-1.4.html
+[1.5]: https://blog.rust-lang.org/2015/12/10/Rust-1.5.html
+[1.6]: https://blog.rust-lang.org/2016/01/21/Rust-1.6.html
+[1.10]: https://blog.rust-lang.org/2016/07/07/Rust-1.10.html
+[1.12]: https://blog.rust-lang.org/2016/09/29/Rust-1.12.html
+[1.13]: https://blog.rust-lang.org/2016/11/10/Rust-1.13.html
+[1.14]: https://blog.rust-lang.org/2016/12/22/Rust-1.14.html
+[1.15]: https://blog.rust-lang.org/2017/02/02/Rust-1.15.html
+[1.17]: https://blog.rust-lang.org/2017/04/27/Rust-1.17.html
+[1.20]: https://blog.rust-lang.org/2017/08/31/Rust-1.20.html
+[1.24]: https://blog.rust-lang.org/2018/02/15/Rust-1.24.html
+[1.26]: https://blog.rust-lang.org/2018/05/10/Rust-1.26.html
+[1.28]: https://blog.rust-lang.org/2018/08/02/Rust-1.28.html
+[1.31]: https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html
+[1.34]: https://blog.rust-lang.org/2019/04/11/Rust-1.34.0.html
+[1.39]: https://blog.rust-lang.org/2019/11/07/Rust-1.39.0.html
+
+## Error Diagnostics
+
+One thing that we haven't mentioned much is how much Rust's error messages and
+diagnostics have improved since 1.0. But they have improved so much over these
+past five years, looking at older error messages now feels like looking a
+different language.
+
+We’ve highlighted a couple of examples that best showcase just how much we’ve
+improved showing users where they made mistakes and importantly help them
+understand why it doesn’t work and teach them how they can fix it.
+
+##### First Example (Traits)
+```rust
+use std::io::Write;
+
+fn trait_obj(w: &Write) {
+    generic(w);
+}
+
+fn generic<W: Write>(_w: &W) {}
+```
+
+
+<details>
+ <summary>1.2.0 Error Message</summary>
+
+```
+   Compiling error-messages v0.1.0 (file:///Users/usr/src/rust/error-messages)
+src/lib.rs:6:5: 6:12 error: the trait `core::marker::Sized` is not implemented for the type `std::io::Write` [E0277]
+src/lib.rs:6     generic(w);
+                 ^~~~~~~
+src/lib.rs:6:5: 6:12 note: `std::io::Write` does not have a constant size known at compile-time
+src/lib.rs:6     generic(w);
+                 ^~~~~~~
+error: aborting due to previous error
+Could not compile `error-messages`.
+
+To learn more, run the command again with --verbose.
+```
+
+</details>
+
+![A terminal screenshot of the 1.2.0 error message.][trait-error-1.2.0]
```

Comment:
These pictures 404 in [the preview](https://github.com/rust-lang/blog.rust-lang.org/blob/a0dab5676443cfe8137c1ebd5ad76a091ade2570/posts/2020-05-15-five-years-of-rust.md), but maybe that is expected?

Reply:
Yep, it’s a limitation since the current structure requires them in a separate folder that wouldn’t be reachable in the pre-rendered markdown. You can see the images in the top post behind a disclosure though.